add Chart2Music keyboard and sound accessibility features

aliwelchoo opened the pull request:
Use data and layout to call the chart2music API inside makePlot.

This pull request is for scatter plots only

Demo on codepen: https://codepen.io/marthacryan/pen/wvVMrEe

Commit:
initial commit, single scatter plot

Commit:
single scatter plot

Commit:
options and info

Commit:
data and trace labels

Commit:
false default

Commit:
remove debugging prints

Commit:
use index if no x provided

Commit:
Lib requirement

Commit:
layout destructuring

Commit:
layout defaults

Commit:
destructure trace

Commit:
test sourcetype module

aliwelchoo commented:
- Closed captions element is required to be visible on the screen to function so need to find a way to integrate it more cleanly
- Need to rebase

Commit:
remove sourcetype

alexcjohnson commented on src/plot_api/accessibility.js:
```diff
@@ -0,0 +1,71 @@
+                pointNumber: index
+            }])
+            },
+            ...gd._context.chart2musicOptions
+        },
+        info: gd._context.chart2musicInfo
```
Maybe we can combine these into one and [`extendDeep`](https://github.com/plotly/plotly.js/blob/149c895db2acf3f497a84d62adc276c24d37989b/src/lib/extend.js#L24) the config object?

Also I was discussing this PR with @archmoj and he had a good idea, instead of
```js
config: {chart2music: true, chart2musicConfig: {...}}
```
how about we make an `accessibility` object like:
```js
config: {
    accessibility: {
        library: 'chart2music', // easier to extend, but for now chart2music would be the only option, and the default if you make an accessibility object.
        enabled: true, // optional but would allow disabling if necessary
        options: {...}
    }
}
```

aliwelchoo replied:
I like that. That was the idea behind the file naming, makes sense to have the same structure in the config

aliwelchoo replied:
These changes have been made, I didn't use extendDeep (just because I wasn't sure on usage) but options and info are now combined

alexcjohnson commented:
Excited for this! 
FYI @julianna-langston @spmealin - I think we can sort out most of what we need but your feedback is of course welcome!

@archmoj thoughts about packaging? Is a 50kb increase in minified bundle acceptable or should we keep `chart2music` external like `MathJax`? If we did make it external we'd probably want it included by default in plotly.py output, and we could have Dash detect and autoload it, but plain plotly.js users would need to explicitly load it

archmoj commented on package.json:
```diff
@@ -77,6 +77,7 @@
     "@turf/bbox": "^6.4.0",
     "@turf/centroid": "^6.0.2",
     "canvas-fit": "^1.5.0",
+    "chart2music": "^1.11.0",
```
Nice MIT library.
To transpile it to es5, please add it to 
https://github.com/plotly/plotly.js/blob/149c895db2acf3f497a84d62adc276c24d37989b/webpack.config.js#L25

aliwelchoo replied:
Thank you! Knew there'd be something simple I was missing so I could get this into the same style

archmoj commented on src/plot_api/accessibility.js:
```diff
@@ -0,0 +1,71 @@
+"use strict";
+import c2mChart from "chart2music";
```
To pass current es5 build tests, this file should be written in es5 style similar other files to `src`.

alexcjohnson replied:
Do we still need to be writing in es5? Won't webpack take that and give us the desired output? I'd love it if we can start accepting (and eventually converting existing code to) more modern syntax.

archmoj replied:
For es modules, we have a config rule in place.
I did try to add a rule to transpile this file during build processes; but `npm run build` wasn't successful in the past hour.
If one could help with setting up the webpack config, it would be great :pray:

aliwelchoo replied:
Thank you for these tips again! I've made some attempts at this (putting chart2music and/or accessibility.js into the webpack config but struggling with "Cannot find module 'chart2music'" whenever I use require instead of import. I think I'm still missing something

archmoj commented on src/plot_api/accessibility.js:
```diff
@@ -0,0 +1,71 @@
+        var {type, x = [], y = [], name = i, text = []} = trace;
+        if(type === 'scatter') {
+            var traceData = [];
+            if ('y' in trace) {
```
In scatter traces with `orientation: 'h'`, we should use `trace.x` instead of `trace.y`.

archmoj commented on src/plot_api/accessibility.js:
```diff
@@ -0,0 +1,71 @@
+        }
+        else {
+            // 'Accessibility not implemented for trace type: ' + trace.type
+            return;
```
Perhaps we should use `continue` instead of `return` here.

alexcjohnson replied:
Yes I think so. There's some overview text c2m reads out when you first focus on the chart, right? Perhaps we can have any skipped traces briefly mentioned there?

aliwelchoo replied:
I can look into this. Need to look into where that closed captions text element goes in general, not sure about outputting our own messages into it, it's more than just the overview at the start, also reads off the data as you navigate it.

But this general point is a good one, ideally want to be able to support more than just scatter traces and want to support figures containing unsupported traces alongside supported ones. 

Then there's also the opportunity for a config option to say which traces should be in accessibility mode

alexcjohnson replied:
@julianna-langston that would be great! I think that's going to be really important for this project, so that plotly.js and Dash users who care about this can turn on chart2music in all of their charts, knowing that the result will not always be complete but will always be as useful as it can be and will tell them what they're missing. And then over time as both the plotly.js and the chart2music sides add more features, existing applications will automatically improve.

julianna-langston replied:
Chart2Music v1.12 now has an `unsupported` chart type. [Examples and documentation](https://chart2music.com/docs/API/Developer#unsupported-groups-of-data)

julianna-langston commented on src/plot_api/accessibility.js:
```diff
@@ -0,0 +1,71 @@
+    
+    c2mChart({
+        title: title_text,
+        type: "line",
```
If this PR is for scatter plots, you probably want to have `type: "scatter"`.

Also note that, if you want to support more chart types, chart2music supports an array. So, if you have 2 traces, 1 scatter and 1 line, you could convert that to chart2music's `type: ["line", "scatter"]`.

aliwelchoo replied:
Thank you. I think I'll end up extending the scope to as many types as we can

Coding-with-Adam replied:
Thank you for supporting our project, @julianna-langston

Commit:
into es5

Commit:
include mjs files

Commit:
transpile accessibility

Commit:
accessibility in config

aliwelchoo commented on webpack.config.js:
```diff
@@ -22,7 +22,7 @@ module.exports = {
         }, {
             test: /\.js$/,
-            include: /node_modules[\\\/](buffer|d3-color|d3-interpolate|is-mobile)[\\\/]/,
+            include: /node_modules[\\\/](buffer|d3-color|d3-interpolate|is-mobile|chart2music)[\\\/]/,
```
This change unfortunately doesn't do anything to resolve the build check issues. Is there something else needed?

archmoj replied:
This still helps converting `chart2music` dependency to es5. So let's keep it here.
The build issue you are mentioning is related to new file `src/plot_api/accessibility.js` being written in es6.
@alexcjohnson wondering what `webpack` configuration we used in other libraries to bundle es6 files in `src`?

aliwelchoo replied:
When I write that file in es5 using require instead of import, it can't find the chart2music module. I have tried putting type 'module' in the package.json. I think maybe something to do with chart2music not being a .mjs file

ayjayt committed:
Merge branch 'master' into chart2music

ayjayt committed:
Remove special rules from webpack.config:

I agree with the contributor's choice of dialetc over plotly's default.

However, I think adding piece-meal rules to change how stuff is built
will ultimately make it more complicated to refactor, maintain, and read
in the future.

The current dialect of javascript (enforced primarily by the linter at
least in the pull-request stage), is limiting and inconvenient.
Migrating to a more modern javascript, I believe, requires a broader
strategy.

ayjayt committed:
Reorder declarations to make a bit more readable

ayjayt committed:
Lint

ayjayt committed:
Change import/export scheme to promise

ayjayt committed:
Remove a console.log

ayjayt committed:
Limit chunks to 1

ayjayt commented:
### Where we're stuck

**Warning: These problems will increase in frequency because ESM is highly recommended**

You cannot `require()` (**commonjs**) a `.mjs` (**ESM**) because the `require()` will block and wait for the `.mjs` to load and the `.mjs` may `await` expecting that execution would continue without it.. but a) it won't and b) sync code can't `await`.

Despite that your configuration would nullify that issue, (theory:) the _webpack_ resolver sees you're `require()`ing a `.mjs` and simply fails silently (anticipating above) and decides not to include it in the dependency tree, which is why none of the loaders pick it up to do transpilations. It's just not being included in your input or output.

Furthermore, the _babel_ plugin to do a **ESM**-->**commonjs** transpilation will also fail silently if any of the `imports` of the input code are **ESM** only-- but luckily, `chart2music` has no dependencies at all! But that's not the only obstacle to using _babel_ to transpile it: (theory:) the `package.json` insists `chart2music` is an **ESM** module and the _babel_ plugin doesn't modify that, so _webpack_ still chokes the moment it gets to `package.json` of that module.

### Solutions

1)

There is a `commonjs` dynamic `import()` compatible with **ESM** but the linter is blocking it.  [Using it works!](https://github.com/geopozo/plotly.js/tree/pikul-music-import-promise) but would cause, for the first time, your bundler to output two separate chunks (click and zoom pls):

![image](https://github.com/plotly/plotly.js/assets/30324885/4d83fc2d-760d-4719-abfe-e9be25be4b0b)
(Master branch at bottom)

Since you guys distribute whole files, until you're ready (lets get there soon) for a real refactor and optimization for `plotly` delivery, you might want to tell webpack **no chunking** so it doesn't create separate files (as seen in my branch).

Or, if you were going to include this as a button on the mode bar, you would use dynamic `import()` to fetch the dependency and redo `_doPlot()` on-click.


2)
Running the `babel-cli` and changing the `package.json` to not insist on being **ESM** does work, so it could be included in the module. Could make a pull on the main repo.


Original Dep Tree of Master Branch:
![image](https://github.com/plotly/plotly.js/assets/30324885/0a35498e-15b4-46a1-b1ef-6ae2ab185386)

ayjayt committed:
Upgrade chart2music to support commonjs

ayjayt committed:
Rewrite accessibility in es5

ayjayt commented:
`chart2music` merged in my request to support **commonjs**, so I submitted a [pull request](https://github.com/aliwelchoo/plotly.js/pull/1) to @aliwelchoo that should make this go green.

aliwelchoo committed:
Merge pull request #1 from geopozo/pikul-music-import-promise

Update to your pull-request for making es5/require work.

ayjayt committed:
Fix some undef errors in accessibility

ayjayt committed:
update plot-schema diff

ayjayt commented:
Alright, I fixed the obvious stuff (including plot-schema) from those errors, but non-obvious stuff is:

1) What should plotly.Purge() do with the closed-caption div. Should we change to the test to expect an empty closed-caption div?

```
1) should return the graph div in its original state
     Test plot api Plotly.purge
     Error: Expected '<div id="cc" class="closed_captions" aria-live="assertive" role="status" aria-atomic="true" aria-relevant="additions text"></div>' to equal ''.
    at <Jasmine>
    at eval (webpack://plotly.js/./test/jasmine/tests/plot_api_test.js?:2340:30)
```
2) No clue what baseline tests are:

```
domain_refs
Transform failed with error code 525: Cannot read property 'x' of undefined
retry 1 / 2
Transform failed with error code 525: Cannot read property 'x' of undefined
retry 2 / 2
Transform failed with error code 525: Cannot read property 'x' of undefined
```

https://app.circleci.com/pipelines/github/plotly/plotly.js/9547/workflows/fbf28531-254a-4bb8-9722-924609ce9de4

ayjayt committed:
Merge branch 'master' into pikul-music-import-promise

ayjayt committed:
Add config so dev can place closed captions:

Adding a closed caption div to plotly interefered with plotly's
management of it.

We now start by default outside of the generated plotly divs.

But we add options to users can a) generate it with any id/class.

Or specify that it already exists and use an existing element by id.

ayjayt committed:
Merge branch 'chart2music' of https://github.com/aliwelchoo/plotly.js into pikul-music-import-promise

ayjayt committed:
Add more default assurance in accessibility

ayjayt committed:
Lint

ayjayt committed:
Skip empty trace data in accessibility.enable:

Passing named empty arrays to chart2music makes it angry during
validation.

ayjayt committed:
update plot-schema diff

ayjayt committed:
Modify how accessibility treats config vars:

accessibility.js takes configuration variables from `plot_config.js`, modifies
them, and passes them directly to `chart2music`.

The problem is that assigning an object to a new variable doesn't copy
it, so any change to that variable also changes the original object.

Plotly uses changes (or lack thereof) in the config variables to decide if it can
shortcut certain redraw steps. It cannot if config variables change. So
if `accessibility.js` is run on every redraw, the config variables will look
different (even if they are not), and plotly will always take the long
way around during its rerendering.

That being said, this solution here may need more thought. Questions
like: when _should_ chart2music ask plotly to do a full redraw? Should
it ever? Would it do that by changing a config it has access to? Will it
ever change plotly config variables frequently.

ayjayt committed:
Remove jasmine ftest focus

ayjayt committed:
Run schema plot diff

ayjayt committed:
Test if previous commits deprecate hack-fix

ayjayt committed:
Set c2music default to not make its caption div:

I can't think of instance where

1) the caption div wouldn't be specified by the user
2) would maybe do some really fancy interaction w/ the graph
3) be completely invisible because its meant to be read outloud anyway

Either way, the generation strategy needs work.

* Tests now create their closed caption div

ayjayt committed:
update plot-scheme diff

ayjayt committed:
Add closed captions to dev dashboard

ayjayt commented:
I've just made a pull request on this repo:branch w/ about a million commits that makes us go green, but there are issues, let's see:

### 1) The closed caption div. By default it doesn't generate automatically.
  -  Tests now create a closed caption div in their `createGraphDiv()` in `assets/`
  -  Dev dashboard now comes with a nice one stock.

If it's going to generate automatically, and INSIDE the graph div, someone needs to quiet down all the tests and functions that are surprised to see a new neighbor.

@julianna-langston Does it  (closed-caption div) even need to be visible to work?  :-)

### 2) Where to rerender C2M?

Plotly has several possible renders eg. `transitionFromReact()`, `newPlot()`. `accessibility.js`) does everything in  `_doPlot()`. `accessibility.js` should split into 2: a) set up basic config b) (re)count data). Or what? @archmoj @alexcjohnson I have no idea who this question is for. 

### 3) Large Accessibility Concerns

I think probably that when someone key-tabs into plotly, it should focus first to a button on the mode bar (accompanied by a temporary closed caption?) that can activate accessibility mode. I don't think closed captions should just _appear_ anytime someone clicks the graph with the mouse. I'm assuming users will generally being key-tabbing into it and so that makes a good entry point for enabling accessibility mode. Of course, other uses can click the mode button and turn it on and off. Opinions? @julianna-langston... I'm tagging you because I have no experience with screen readers and I'm not sure what the user experience should be like.

julianna-langston commented:
> Does it (closed-caption div) even need to be visible to work? :-)
You can position the closed caption off screen, or you can use `visibility:hidden`, but you can't use `display: none`.

There was a few months there where either Firefox or NVDA wasn't recognizing closed captions that were positioned off screen, but I haven't seen that problem in over a year. I would feel comfortable with simply moving the box off screen with something like this:

```css
#cc {
  position: absolute;
  top: -1000px;
  left: -1000px;
}
```

>  I don't think closed captions should just appear anytime someone clicks the graph with the mouse
If you position the closed caption off screen, you can support screen reader users, without sighted users ever knowing.

> I'm assuming users will generally being key-tabbing into it and so that makes a good entry point for enabling accessibility mode.
This is a fair assumption for sighted, keyboard-only users.

However, screen reader users frequently navigate a page with shortcut keys that come with the screen reader. Many of these shortcuts will allow them to jump directly to a type of element. For example, you can navigate by headings, links, buttons, or images, among other options. Notably, if the poltly graphs have `role="img"` associated with them, a screen reader user could get that graph in their list of images, and then be able to jump directly to it. 

While it's obviously best for screen reader support to always be on, it's still perfectly reasonable to put accessibility behind a button, as long as that button is discoverable. The button can be considered discoverable if:
1. It's located nearby *in the DOM* - being visually nearby doesn't help. 
2. Clearly labeled, such as saying "Turn on accessibility mode" or "Turn on screen reader support".

I hope that's helpful. Please let me know if you have any other questions!

ayjayt commented:
thanks @julianna-langston, I take it the `<div>` must have `id=cc`, or would other `id` work?

### Todo:

- [ ] sensible defaults for CC generation (off screen)
- [ ] test user-flow with actual screen reader (always on? toggleable via modebar? modebar is accessible?)
- [ ] Add `role=img`?
- [ ] refactor to make for better rendering (see note)
- [ ] refactor to deal with subplots and other axes
- [ ] add reasonable data structures:
  - [ ] scattergl
  - [ ] category?
  - [ ] i dunno
### Proposed Structure:

* Two `c2m` functions: `accessibility.fullRedraw()` `accessibility.reprocessData()`

  * `accessibility.fullRedraw()` will create a `c2m` instance in `gd._context._c2m`

  * `accessibility.reprocessData()` if traces are added or removed, I suppose (what if user zooms? can we get visible?)

* Look into how accessibility configs (stored in `_context`) are changed in run time- if through a `c2m` api, `c2m` must communicate config change to `gd` to avoid `transitionFromReact()` and prompt whole-redraw. But if `c2m` is using its `options` object for example as short term memory, it shouldn't trigger a `configFlag` change. So `c2m`/`accessibility.js` need to understand when they should trigger a full redraw of the plot. Likewise, `gd` will `seq.push()` the appropriate `accessibility.` function during its many render functions but have to decide which is called and where. 

**Future**
This will play into my thoughts about adding optional down-sampling later this year, since listening to over-renderer data would be worse than seeing it. Interestingly, a competitor to dash who _uses_ plotly implemented some algorithms and published about it, but didn't push it into plotly. Forget who off the top of my head.

julianna-langston commented:
> I take it the <div> must have id=cc, or would other id work?

You can pass any kind of DIV you want to the c2mChart's `cc` option. The DIV can have any value for ID, or even not define an ID, and it will still work.

aliwelchoo committed:
Merge pull request #2 from geopozo/pikul-music-import-promise

FIX and Add options to chart2music plotly integration

ayjayt committed:
Merge branch 'master' into chart2music

ayjayt committed:
Merge branch 'master' into chart2music

Coding-with-Adam commented:
Thank you for working on this project, everyone. We're trying to figure out a way to support ES6 in our code base. We should have an update about this soon, which will allow us to make more progress implementing Chart2Music.

ayjayt commented:
yeah rollup can do it out of box tbh.

alexcjohnson commented:
@ayjayt FYI we're working on transitioning this project from `webpack` to `esbuild`. `webpack` can of course be made to handle modern syntax and output to any feature set we want, but this project has some odd quirks particularly around the WebGL code that have made it tricky for us to get right, and anyway we're provisionally seeing some great benefits to switching to `esbuild`. So hang on just a bit, hopefully next week we'll have this sorted out and be able to move forward 🙏

ayjayt commented:
ESM/CJS preferences aside, here's new todo:

- [ ] merge with main and make go green again
### **Little Stuff**
- [ ] sensible defaults for CC generation (off screen)
- [ ] use `Lib.getGraphDiv()`
- [ ] Add `role=img`?
- [ ] test user-flow with actual screen reader (always on? toggleable via modebar? modebar is accessible?)
### **Big Stuff**
- [ ] c2m should copy default config
- [ ] c2m should store its handler in a hidden _context variable (relook at how config spurs redraw)
- [ ] c2m should also store it's copied config in hidden _context variables
- [ ] When c2m is reinitialized in the same plot (for whatever reason), use the stored, not the default, so that any config changed at runtime is maintained
- [ ] c2m will never provoke plotly to redraw, all data flows from plotly to c2m
- [ ] c2m data can be redrawn on the zoom/pan callback (i think there is a callback), although I'm not quite sure how a visually impaired person would use zoom/pan yet. make sure we get current selected data point and make that same current selected data point after change? unless out of frame? hmm
- [ ] follow instructions in c2m documentation about supporting hover
- [ ] see if maybe c2m left/right hotkeys can cause panning when we get to end of data
- [ ] add ability to skip notes
- [ ] add modebar button 
- [ ] all conversions from plotly-->c2md data structures will be modularized as codecs
- [ ] failed conversions will have a debug option to dump information so new codecs can be written
- [ ] add support to manually switch between discrete/continuous mode in c2m, perhaps as precursor to intelligently deciding.

julianna-langston commented:
I'm happy to help with the Chart2Music side! Please let me know what I can do to facilitate knowledge transfer, or if there are specific features you need from C2M.

ayjayt commented:
Hi @julianna-langston ! I think that the c2m library is fine, when I say "c2m" side I'm just referring to the plotly module that implements it.

ayjayt committed:
Merge branch 'master' into chart2music

ayjayt committed:
Refactor accessibility/c2m:

1) Since accessibility.js will support multiple accessibility libraries,
   there is now a C2M library that contains the API for initializing C2M
   to plotly charts.
2) The initialization process is split into it's various phases.
3) Trace types now have specific codecs to map plotly-C2M. See the
   README in each file.
4) xy-scatter codec was refactored a bit to not assume the x-y is
   domain-range.
5) All config variables are now copied from defaults so C2M doesn't mess
   with plotly's config obect.

ayjayt commented:
I've refactored this branch to give `accessibility.js` a proper API.

Since `accessibility.js` is set up to support multiple accessibility libraries, `accessibility.js` no longer contains `c2m`-specific logic, but instead calls a `c2m` library internal to plotly (a wrapper for the actual `chart2music` library).

This to-do list builds on the above:

- [ ] The code is now littered with TODO's that need to be addressed, so these two lists are not comprehensive!
- [ ] Need to create *_codec.model files which will serve as guides for future codecs
- [ ] Need to reread C2M docs about hiding/showing traces
- [ ] Need to consider live data
- [ ] Need to get plotly expert to tell where plotly should call full-scale reinit, or just reinit data.
- [ ] Need to consider data width

If you're interested in the development, consider the [development docs](https://github.com/plotly/plotly.js/blob/0c434076ff41708a827e5fda3caf798f42e3a622/src/accessibility/c2m/README.md).

ayjayt committed:
Lint

ayjayt committed:
Fix import path

Coding-with-Adam commented:
hi @ayjayt 
Thank you for all your work on this. Can I please have your email or linkedin? I'd like to share more side thoughts with you. 
adam@plot.ly

Thanks,

ayjayt commented:
> hi @ayjayt Thank you for all your work on this. Can I please have your email or linkedin? I'd like to share more side thoughts with you. [adam@plot.ly](mailto:adam@plot.ly)
> 
> Thanks,
Sure,

https://www.linkedin.com/in/ajpikul/
ajpikul@gmail.com

Although this thread is a great place to have the discussing as well!

Coding-with-Adam commented:
> Although this thread is a great place to have the discussing as well!

Of course, I plan to discuss issues related to this topic only here. 
Just wanted to share a personal note via email. Thanks for sharing.

archmoj committed:
Merge remote-tracking branch 'origin/build-with-esbuild' into chart2music_dev

archmoj commented:
Alright, I opened #6909 and merge it into [chart2music_dev](https://github.com/plotly/plotly.js/tree/chart2music_dev) branch. All the tests are green there. So let's pull/merge that here.

ayjayt commented:
thanks @archmoj!

ayjayt committed:
Merge branch 'master' into chart2music

ayjayt commented:
Todo 1: https://github.com/plotly/plotly.js/pull/6680#issuecomment-1962404214
Todo 2: https://github.com/plotly/plotly.js/pull/6680#issuecomment-1968198454

just copying this here because it all got lost with the big merge

ayjayt commented:
https://github.com/julianna-langston/chart2music/pull/416

ayjayt commented:
This is on track! When the infrastructure is done (probably next weekend), it will be ready for systematic integration with all of plotly's plot types. I will get us started with 2 or 3 (aliwelchoo already did a lot of linear), but I'm not as familiar with all of plotly's plot types as some of you so I will write documentation to make it easy, organized, and systematic to:
1. discover which of plotly's graph types aren't supported yet
2. write the pursuant support

Coding-with-Adam commented:
Thank you so much with your support on this, @ayjayt

archmoj commented:
We fixed a problem on CircleCI (#6935). 
Please pull `upstream/master` and merge it into this branch.

ayjayt committed:
Merge branch 'master' into chart2music

ayjayt committed:
Merge branch 'master' into chart2music

ayjayt commented:
I'm back 😎, lets see what we can´t finish today :-)

marthacryan commented:
@aliwelchoo @ayjayt @julianna-langston Hi! I'm trying to pick up this PR and help get it merged. I'd love to have a call to discuss what's left on this PR and how I can help. Let me know!

archmoj commented:
Thanks @marthacryan for the note.
There are some conflicts that must to be resolved.
One should fetch `upstream/master` and merge it into this branch.
I could help with that as well.
Would you like to discuss it on a call now?

marthacryan committed:
Merge with master

marthacryan commented on test/plot-schema.json:
```diff
@@ -104,6 +104,21 @@
+   "description": "Accessibility options: which library to use; whether to enable, options to pass to the library, info to pass to the library, closedCaptions to control how plotly renders the closed-captions element. chart2music is supported and options here include Options and Info from https://www.chart2music.com/docs/API/Config. ",
+   "dflt": {
+    "closedCaptions": {
+     "elClassname": "c2m-plotly-closed_captions",
```
We might want to discuss whether we want to use the name of the tool in the schema.

ayjayt commented:
I think I can square up a couple days to work on this fulltime maybe 27th-30th of september...

Send me an e-mail and I'm happy to give out my whatsapp to make communication easier.

marthacryan commented on devtools/test_dashboard/index-mathjax3.html:
```diff
@@ -24,6 +24,10 @@ <h4>no MathJax: Apple: $2, Orange: $3</h4>
     <div id="graph"></div>
   </div>
   <div id="snapshot"></div>
+  <div style="position:absolute; border-top: .3em solid #eee; bottom: 0px; left: 0; width: 100%">
```
I'm not sure if we want this to be included in every test dashboard by default. Maybe we could add a special flag for that?

marthacryan commented on devtools/test_dashboard/index-mathjax3.html:
```diff
@@ -24,6 +24,10 @@ <h4>no MathJax: Apple: $2, Orange: $3</h4>
   <div id="snapshot"></div>
+  <div style="position:absolute; border-top: .3em solid #eee; bottom: 0px; left: 0; width: 100%">
+    <div style="margin-top: -1.2em; padding-left: 3px; font-weight: bold; color: #555; width:100%;">Closed Captions</div>
+    <div id="c2m-plotly-cc" style="width: 100%; height: 40px; padding: 18px 0 0 15px; color: #333">Test</div>
```
We could have something more informative than "Test" here as the default value.

marthacryan commented on src/accessibility/README.md:
```diff
@@ -0,0 +1,3 @@
+# accesibility
```
Accessibility spelling is incorrect

marthacryan commented on test/plot-schema.json:
```diff
@@ -104,6 +104,21 @@
   }
  },
  "config": {
+  "accessibility": {
```
Maybe we could name this config "sonification" instead of accessibility? Or is this more than sonification?

marthacryan commented:
Could someone confirm for me that I'm understanding this correctly: this PR includes the addition of _both_ sonification **and** closed captions? Are they related other than both being accessibility features?

ayjayt commented:
This PR is an integration of all of c2ms features with plotly. c2m provides
both sonification and closed captions. At the minimum, closed captions are
needed to instruct users as to c2ms interface, so c2m provides those
functions.

El mar, 17 de sept de 2024, 9:29 a. m., Martha Cryan <
***@***.***> escribió:

> Could someone confirm for me that I'm understanding this correctly: this
> PR includes the addition of *both* sonification *and* closed captions?
> Are they related other than both being accessibility features?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/plotly/plotly.js/pull/6680#issuecomment-2355803815>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHHLRFL6NB7HGL5QCZ2SRP3ZXAVCNAVCNFSM6AAAAAA2TEO5UGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNJVHAYDGOBRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

marthacryan commented:
When I tested this out, I noticed that the keybindings listed are incorrect for mac.

marthacryan commented:
Before merging, we should check that this works in jupyterlab. I've seen issues with tab traps in the notebook, so this being something that relies on the user tabbing it into focus could be an issue. Not that that's chart2music's fault! But it's worth considering workarounds, etc.

ayjayt commented:
Is jupyter lab an accessible environment? I'm honestly asking, I have no
idea. If it's not, there won't be a strong use-case for c2m in jupyter lab
until it is. I could definitely see published notebooks being something
worth investigating though!

On Tue, Sep 17, 2024 at 10:07 AM Martha Cryan ***@***.***>
wrote:

> I also have had issues with the codepens from that repo sometimes not
> allowing me to tab the chart into focus. Is there a way around that?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/plotly/plotly.js/pull/6680#issuecomment-2355938712>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHHLRFPHP4CYD6PQP4TBQDTZXAZRLAVCNFSM6AAAAAA2TEO5UGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNJVHEZTQNZRGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

marthacryan commented on src/plot_api/plot_config.js:
```diff
@@ -471,7 +471,21 @@ var configAttributes = {
+            enabled: true,
+            options: {},
+            info: {},
+            closedCaptions: {generate: false, elId: 'c2m-plotly-cc', elClassname: 'c2m-plotly-closed_captions'}
```
I feel like all of these options are kind of overkill. Why do we need to specify both id and classname? And why are we making it an option that _we_ generate the closed caption div? I think we should just simplify this into one field called `closedCaptionsElId` and just assume users will make that div. I don't want to overcomplicate this

marthacryan replied:
Looking into this more I actually think we can just keep these in.

marthacryan committed:
Change name of config to sonification, clean up directory

marthacryan committed:
Spelling

marthacryan committed:
fix directory name

marthacryan committed:
Reset package-lock

marthacryan committed:
Update package-lock.json

marthacryan committed:
Reset tests

marthacryan committed:
Update auto-generated schema

ayjayt commented:
Here to provide any technical support I can, just @ me

marthacryan committed:
Rename more stuff to sonification and clean up contributing

marthacryan committed:
Update filepath for new location of enable_sonification.js

marthacryan committed:
Code readability improvements

marthacryan commented on src/plot_api/plot_api.js:
```diff
@@ -381,10 +382,21 @@ function _doPlot(gd, data, layout, config) {
+        Plots.doAutoMargin
     );
 
+    function saveRangeInitialForInsideTickLabels(gd) {
```
I'm not clear on why this was added? How does it relate to the chart2music? @ayjayt

marthacryan replied:
Actually, looking at the git blame it looks like this function was added in 2020 by @archmoj? I'm not sure why it's even showing up in this diff

archmoj replied:
Could you please point to the commit that added this?

archmoj replied:
If we don't have this on master, this function should be removed. It might be the result of a merge conflict going wrong.

marthacryan replied:
There are two commits for this function: 
* https://github.com/plotly/plotly.js/commit/5ca95d3794159ace3eb623ea1a7f62a3f6971106
* https://github.com/plotly/plotly.js/commit/445eee4ee21cadef37981e57db63619ef6d12b66

archmoj replied:
This function is not relant to this PR.
Please remove it.

marthacryan commented on src/accessibility/sonification/index.js:
```diff
@@ -0,0 +1,104 @@
+
+function initC2M(gd, defaultConfig) {
+    // TODO: should this be Lib.getGraphDiv()?
+    // TODO what is there besides a fullReset?
```
What does this comment mean? @ayjayt

ayjayt replied:
Hi sorry! I just fixed my github notifications today

Line 10: there is a function `Lib.getGraphDiv()` that seems to coerce `gd` to be a "real" `gd` but I'm not sure when its needed.
Line 20: Me not knowing the different "levels" of plotly's rendering functions (fullReset being a full-rerender, but there are shortcuts) had previously caused bugs, so my basic question is "When do we actually need to re-initialize C2M besides a fullReset?"

alexcjohnson replied:
`getGraphDiv` is for functions to be called by end users, so that they can pass in the ID of their graph div as a string, as an alternative to passing in the DOM element directly. Functions called only by our code do not need this step. Which kind is `initC2M`?

marthacryan commented on src/accessibility/sonification/index.js:
```diff
@@ -0,0 +1,104 @@
+var codecs = require('./all_codecs').codecs;
+
+function initC2M(gd, defaultConfig) {
+    // TODO: should this be Lib.getGraphDiv()?
```
@archmoj or @alexcjohnson might be better equipped to answer this?

archmoj replied:
Passing `gd` as an argument to this function (as implemented) is better that getting it from its id using `Lib.getGraphDiv`.

marthacryan committed:
Add baseline for new mock

marthacryan committed:
Remove changes in existing mock

marthacryan committed:
Update baseline

marthacryan committed:
Remove function that was introduced through incorrect merge

marthacryan commented on test/jasmine/assets/create_graph_div.js:
```diff
@@ -5,6 +5,11 @@ module.exports = function createGraphDiv() {
     gd.id = 'graph';
     document.body.appendChild(gd);
 
+    var closedCaptions = document.createElement('div');
```
I don't think we need this since we're not editing the jasmine tests in this PR?

marthacryan committed:
Undo changes to jasmine tests

marthacryan commented on src/accessibility/sonification/index.js:
```diff
@@ -0,0 +1,104 @@
+    var ccElement = initClosedCaptionDiv(gd, c2mContext.ccOptions);
+
+    // TODO:
+    // I believe that title OR title.text could contain the information needed
```
@archmoj Is that accurate that title or title.text could contain the information needed?

archmoj replied:
The preferred way of setting a title is via `title.text` not `title`. 
Setting `title` to a string is discouraged(& deprecated) but may still widely used in our codes.
We may handle case of `title` in case title is set to a string instead of an object.

marthacryan replied:
We should drop support for that!

archmoj commented on src/accessibility/sonification/index.js:
```diff
@@ -0,0 +1,104 @@
+    // I'd rather the callback be given its c2m handler which
+    // could store extra data. Or bind c2mContext as `this`.
+
+    var titleText = gd._fullLayout.title?.text ?? 'Chart';
```
Wondering if we could/should use other default e.g. `Plotly chart` or something else?

marthacryan replied:
@ndrezn Thoughts?

ndrezn replied:
I'm fine with just "Chart"

marthacryan committed:
Revert "Code readability improvements"

This reverts commit 129eb3b57779d2a01aa3976a92e3884356242a12.

marthacryan committed:
Fix syntax

marthacryan committed:
remove comment

marthacryan committed:
Remove todo

marthacryan committed:
Schema update

marthacryan committed:
Merge branch 'master' of github.com:plotly/plotly.js into chart2music

marthacryan committed:
Remove accessibility wrapper directory

marthacryan committed:
Fix filepaths

marthacryan committed:
Code cleanup

marthacryan committed:
Code cleanup

marthacryan committed:
Code cleanup

archmoj commented on src/plot_api/plot_api.js:
```diff
@@ -381,10 +382,14 @@ function _doPlot(gd, data, layout, config) {
+        Plots.doAutoMargin
     );
 
+    if(gd._context.sonification.enabled) seq.push(sonification.enable_sonification);
```
IS `sonification` key always present?

archmoj commented on src/sonification/index.js:
```diff
@@ -0,0 +1,103 @@
+        closedCaptions.className = config.elClassname;
+        gd.parentNode.insertBefore(closedCaptions, gd.nextSibling); // this really might not work
+        return closedCaptions;
+    } else {
+        return document.getElementById(config.elId);
+    }
```
We could simply return and drop else.
```suggestion
    }
    return document.getElementById(config.elId);
```

archmoj commented on src/sonification/xy_scatter_codec.js:
```diff
@@ -0,0 +1,25 @@
+    var x = trace.x && trace.x.length !== 0 ? trace.x : [];
+    var y = trace.y && trace.y.length !== 0 ? trace.y : [];
+
+    for(var p = 0; p < Math.max(x.length, y.length); p++) {
```
Let's put `Math.max(x.length, y.length)` in a variable before the loop and use it inside the loop.

archmoj commented on src/sonification/xy_scatter_codec.js:
```diff
@@ -0,0 +1,25 @@
+// This codec serves for one x axis and one y axis
+
+function test(trace) {
+    if (!trace || !trace.type || trace.type !== 'scatter') return null;
```
Does it for `scattergl`?

marthacryan committed:
Fix trace bug

marthacryan commented:
Last commit fixes #7193

ayjayt committed:
Merge branch 'master' into chart2music
